### DISCUSSION TITLE
reduce Archive object memory usage

### DISCUSSION THREAD

**Discussion by YYTB** [pymupdf/PyMuPDF#3570] [Q&A] [answered]
when i use my own font file to insert_htmlbox like the link below. but if document have many pages and i need do the insert_htmlbox on every page in a for loop, memory usage increse really fast, and finally cause the program crash. i assume the Archive object is the reason, beacause when i dont use my own font, there is no problem.
how to avoid this?
https://pymupdf.readthedocs.io/en/latest/recipes-text.html#how-to-specify-your-own-fonts
```python
doc = pymupdf.open("test.pdf")
arch = pymupdf.Archive("./fonts")
css = """
@font-face {font-family: user; src: url(SourceHanSrif-Bold.otf);}
* {font-family: user, sans-serif}
"""
for page in doc:
    page.insert_htmlbox(rect, text, css=css, archive=arch)
doc.subset_fonts(verbose=True)
```

**Comment by JorjMcKie** [comment]
The reason is probably the crazy size of that font (24 MB), not the archive.
You can try and force MuPDF to empty its cache after each page:
```python
for page in doc:
    page.insert_htmlbox(rect, text, css=css, archive=arch)
    page = None
    pymupdf.TOOLS.store_shrink(100)
```
If this does not help and the inserted text is the same on each page, there may be other options for reducing memory usage and also speeding up things.

**Reply by YYTB** [reply]
yes, god damn cjk language! LOL. actually i already try .store_shrink(100), it won't help.
maybe i need use insert textbox instead, or just use default font of insert_htmlbox.
thank you very much.

**Reply by JorjMcKie** [reply, marked as answer]
Just tested my own suggestion: **_It will not help!_**

So, depending on the assumption that you indeed **_are_** inserting the same text box on each page, here is a suggestion.

1. Make a temporary PDF with one page having the dimension of the rectangle.
2. `insert_htmlbox` on that temporary page
3. For the pages in your real PDF, execute `page.show_pdf_page(rect, temp, 0)`

This will use no additional memory each time and also be **_very much_** faster!
```python
rect = pymupdf.Rect(100, 100, 300, 300)

# Insert the text on a temporary PDF page with suitable size
temp = pymupdf.open()
tpage = temp.new_page(width=rect.width, height=rect.height)
tpage.insert_htmlbox(tpage.rect, text, css=css, archive=arch)

doc = pymupdf.open(...)  # the real PDF
for page in doc:
    page.show_pdf_page(rect, temp, 0)
```
There still is the following flexibility with this approach:
* The rectangle on the target pages need not have the same position on each page
* The rectangle can be rotated, and also be scaled (which will of course increase/ decrease the text size)
